agent/qrexec-agent-data.c

```diff
@@ -546,6 +572,226 @@ int process_child_io(libvchan_t *data_vchan,
+    size_t const service_length = delimiter ?
+        (size_t)(delimiter - realcmd) : path_length;
+    size_t const argument_length = delimiter ?
+        path_length - service_length - 1 : (size_t)-1;
+#define FAIL(msg, err) do { \
+    fputs((msg "\n"), stderr); \
+    errno = (0 ? ((void)(msg), 0) : (err)); \
+    return -1; \
+} while (0)
+
+    if (argument_length + 1 > QUBES_MAX_SERVICE_ARG_LEN + 1) {
+        FAIL("Service argument too long", E2BIG);
+    } else if (!service_length) {
+        FAIL("Service path empty", EINVAL);
+    } else if (service_length > QUBES_MAX_SERVICE_NAME_LEN) {
+        FAIL("Service path too long", ENAMETOOLONG);
+    } else if (path_length > QUBES_MAX_SERVICE_DESCRIPTOR_LEN) {
+        assert(false && "impossible");
+        abort();
+    }
+#undef FAIL
+
+    if ((s = socket(AF_UNIX, SOCK_STREAM, 0)) == -1) {
+        perror("socket");
+        return -1;
+    }
+
+    for (unsigned use_bare_path = 0; use_bare_path < 2; ++use_bare_path) {
+        for (size_t i = 0; i < sizeof(paths)/sizeof(paths[0]); ++i) {
+            char buffer[QUBES_SOCKADDR_UN_MAX_PATH_LEN];
+            size_t const directory_length = paths[i].length;
+            assert(sizeof buffer > path_length);
+            assert(sizeof buffer - path_length > directory_length);
+
+            // The total size of the path (not including NUL terminator).
+            size_t const total_path_length = directory_length +
+                (use_bare_path ? service_length : path_length);
+            memcpy(buffer, paths[i].string, directory_length);
+            memcpy(buffer + directory_length, realcmd, path_length);
+            buffer[total_path_length] = '\0';
+
+            if (!qubes_connect(s, buffer, total_path_length)) {
+                *stdout_fd = *stdin_fd = s;
+                *stderr_fd = -1;
+                *pid = -1;
+                set_nonblock(s);
+                return 0;
+            }
+            switch (errno) {
+            // These cannot happen
+            case EFAULT:       // all of our parameters are in valid memory
+            case EINVAL:       // we passed valid parameters
+            case ENOTSOCK:     // ditto
+            case ENETUNREACH:  // cannot happen for AF_UNIX
+            case EADDRNOTAVAIL:// cannot happen for AF_UNIX
+            case EINPROGRESS:  // this socket is blocking
+            case EBADF:        // `s` was created by a call to `socket()`
+            case EAFNOSUPPORT: // the kernel supports AF_UNIX
+            case EALREADY:     // ditto
+                abort();
+            // These should not happen
+            case EINTR:
+                // Interrupted by a signal - retry
+                --i; // to retry loop iteration
+                continue;
+            case ENOENT:
+            case ENOTDIR:
+                // These errors could also happen with `execve()`
+                break;
+            case EPROTOTYPE:
+            case ETIMEDOUT:
+            case EACCES:
```

Comment:
A regular file not owned by the user will make `connect` return `EACCES`, so this needs to be handled similar to `ECONNREFUSED`.

@marmarek/@DemiMarie: AFAIU you agreed on the current scheme (try connect /usr/local/etc/qubes-rpc/$x, if refused then exec, then try the same with /etc). Isn't it a bit strange to but the sockets beside the configuration files? What do you think about connect to /run/qubes-rpc-sockets/$x then exec /usr/local/etc/qubes-rpc/$x then exec /etc/qubes-rpc/$x?

Reply:
The use case I had in mind is to place symlinks to sockets (possibly living in `/run`) there, not necessary sockets directly (but still allowed). I think we should document this recommendation.
The reasoning is this: have a clear indicator how the service is really executed. Having separate directory just for sockets (in `/run`) and separate for executables means potentially a different thing being executed depending on the handling service running state. This could also be caused by some race condition on startup, or crashing the service.
If you prevent having both executable and socket for the same service by simply sharing the same file path, it won't be an issue.

Arguably, this could still be the case, if one use `/etc` and `/usr/local/etc`, but it is slightly different as `/usr/local/etc` is meant to override services in `/etc`. And still, one should use symlinks to sockets.

Reply:
Thank you @HW42!  Now fixed.